src/utils/ssh.ts

```diff
@@ -0,0 +1,85 @@
+import { exec } from 'child_process';
+
+import * as vscode from 'vscode';
+import { Logger } from '../logger';
+import { ConfigManager } from '../config-manager';
+
+export class SSH {
+  constructor(private readonly configManager?: ConfigManager) {}
+
+  executeCommand(command: string): Promise<string> {
+    Logger.info(command, 'ssh');
+
+    return new Promise((resolve, reject) => {
+      exec(command, (error, stdout) => {
+        if (error) {
+          reject(error);
+        } else {
+          resolve(stdout);
+        }
+      });
+    });
+  }
+
+  async promptForUsername(hostname: string): Promise<string | undefined> {
+    const username = await vscode.window.showInputBox({
+      prompt: `Please enter the username for host "${hostname}"`,
+    });
+
+    if (username && this.configManager) {
+      this.configManager.setUserForHost(hostname, username);
+    }
+
+    return username;
+  }
+
+  async runCommandAndPromptForUsername(hostname: string, command: string) {
+    const cmd = `ssh ${this.sshHostnameWithUser(hostname)} ${command}`;
+
+    try {
+      const output = await this.executeCommand(cmd);
+      return output;
+    } catch (error: unknown) {
+      if (!(error instanceof Error)) {
+        throw error;
+      }
+
+      if (error.message.includes('Permission denied')) {
+        const username = await this.promptForUsername(hostname);
+
+        if (!username) {
+          const msg = 'Username is required to connect to remote host';
+          vscode.window.showErrorMessage(msg);
+          throw new Error(msg);
+        }
+
+        const cmdWithUser = `ssh ${username}@${hostname} ${command}`;
+
+        try {
+          const output = await this.executeCommand(cmdWithUser);
```

Comment:
For safety and avoiding shell issues, we should always pass the command as a single string, and the arguments as an array, like we do here: https://github.com/tailscale-dev/vscode-tailscale/blob/5f93acc0a8279380d73b4f9460ec606c790b84f3/src/tailscale/cli.ts#L92

Reply:
Yup, done. Though, we do need to send the actual SSH command over the wire with quotes. `ssh foo@bar "date"`